File: seqalt.js
function tokenize(str) {
  const tokens = [];
  for (let i = 0; i < str.length;) {
    let r;
    if (["(", "[", "{"].includes(str[i])) {
      tokens.push({ type: "SequenceStart", string: str[i++] });
    } else if ([")", "]", "}"].includes(str[i])) {
      tokens.push({ type: "SequenceEnd", string: str[i++] });
    } else if (r = str.slice(i).match(/^\d+/)) {
      tokens.push({ type: "Number", string: r[0] });
      i += r[0].length;
    } else if (r = str.slice(i).match(/^"((\\.|[^"])*)"/)) {
      tokens.push({ type: "String", string: r[1].replace(/\\./g, (s) => s[1]) });
      i += r[0].length;
    } else if (r = str.slice(i).match(/^[A-Za-z_]\w*|^[^\w\s\(\)\[\]{}"]+/)) {
      tokens.push({ type: "Symbol", string: r[0] });
      i += r[0].length;
    } else ++i;
  }
  return tokens;
}

function parseSequence(tokens, i, subtype) {
  const exprs = [];
  while (i < tokens.length && tokens[i].type !== "SequenceEnd") {
    const result = parseExpr(tokens, i);
    i = result.i;
    exprs.push(result.expr);
  }
  return { i, expr: { type: "Sequence", subtype, exprs } };
}

function parseExpr(tokens, i) {
  const token = tokens[i];
  if (token.type === "SequenceStart") {
    const result = parseSequence(tokens, i + 1, token.string);
    console.assert(tokens[result.i].type === "SequenceEnd");
    return { i: result.i + 1, expr: result.expr };
  }
  if (token.type === "Number")
    return { i: i + 1, expr: { type: "Number", value: Number(token.string) } };
  if (token.type === "String")
    return { i: i + 1, expr: { type: "String", value: token.string } };
  if (token.type === "Symbol")
    return { i: i + 1, expr: { type: "Symbol", value: token.string } };
}

function parse(tokens) { return parseSequence(tokens, 0).expr; }

function callUserFunc(func, l, r) {
  const env = { __parentEnv: func.env, args: { l, r } };
  if (Array.isArray(r)) func.argNames.forEach((name, i) => env[name] = r[i]);
  else env[func.argNames[0]] = r;
  const val = evalExpr(env, func.expr);
  return val;
}

function callFunc(env, func, l, rExpr) {
  if (typeof func === "function") return func(env, l, rExpr);
  if (typeof func === "object") return callUserFunc(func, l, evalExpr(env, rExpr));
}

function evalSequenceExpr(env, expr) {
  if (expr.exprs.length === 0) return expr.subtype === "[" ? [] : undefined;
  let val = evalExpr(env, expr.exprs[0]);
  if (expr.subtype === "[") val = [val];
  for (let i = 1; i < expr.exprs.length;) {
    const func = evalExpr(env, expr.exprs[i++]);
    const rExpr = expr.exprs[i++] ?? { type: "Null" };
    val = callFunc(env, func, val, rExpr);
  }
  if (expr.subtype === "{" && typeof val === "object") delete val.__isDicUnderConstruction;
  return val;
}

function evalExpr(env, expr) {
  if (expr.type === "Sequence") return evalSequenceExpr(env, expr);
  if (expr.type === "Number") return expr.value;
  if (expr.type === "String") return expr.value;
  if (expr.type === "Symbol") return envVal(env, expr.value);
  if (expr.type === "Null") return undefined;
}

function ownerEnv(env, name) {
  return (!env || name in env) ? env : ownerEnv(env.__parentEnv, name);
}

function envVal(env, name) { return ownerEnv(env, name)[name]; }

function createRootEnv() {
  const rootEnv = {};
  rootEnv["@"] = null;
  rootEnv["rem"] = (env, l, rExpr) => l;
  rootEnv[";"] = (env, l, rExpr) => evalExpr(env, rExpr);
  rootEnv["+"] = (env, l, rExpr) => l + evalExpr(env, rExpr);
  rootEnv["-"] = (env, l, rExpr) => l - evalExpr(env, rExpr);
  rootEnv["*"] = (env, l, rExpr) => l * evalExpr(env, rExpr);
  rootEnv["=="] = (env, l, rExpr) => l === evalExpr(env, rExpr);
  rootEnv["!="] = (env, l, rExpr) => l !== evalExpr(env, rExpr);
  rootEnv["<"] = (env, l, rExpr) => l < evalExpr(env, rExpr);
  rootEnv["<="] = (env, l, rExpr) => l <= evalExpr(env, rExpr);
  rootEnv[">"] = (env, l, rExpr) => l > evalExpr(env, rExpr);
  rootEnv[">="] = (env, l, rExpr) => l >= evalExpr(env, rExpr);
  rootEnv["var"] = (env, l, rExpr) => {
    const name = evalExpr(env, rExpr);
    env[name] = null;
    return name;
  };
  rootEnv["="] = (env, l, rExpr) => {
    if (l instanceof Array) {
      console.assert(l.length === 2);
      const [obj, key] = l;
      return obj[key] = evalExpr(env, rExpr);
    }
    const name = l;
    const e = ownerEnv(env, name) ?? rootEnv;
    return e[name] = evalExpr(env, rExpr);
  };
  rootEnv["."] = (env, l, rExpr) => {
    return l[rExpr.type === "Symbol" ? rExpr.value : evalExpr(env, rExpr)];
  };
  rootEnv[","] = (env, l, rExpr) => {
    if (l?.__isDicUnderConstruction) return { ...l, __tmpKey: evalExpr(env, rExpr) };
    if (l instanceof Array) return [...l, evalExpr(env, rExpr)];
    return [l, evalExpr(env, rExpr)];
  };
  rootEnv["&&"] = (env, l, rExpr) => (l && evalExpr(env, rExpr));
  rootEnv["||"] = (env, l, rExpr) => (l || evalExpr(env, rExpr));
  rootEnv["?"] = (env, l, rExpr) => (Boolean(l) && { __valWhenTrue: evalExpr(env, rExpr) });
  rootEnv[":"] = (env, l, rExpr) => {
    if (l === false) return evalExpr(env, rExpr);
    if (typeof l === "object" && "__valWhenTrue" in l) return l.__valWhenTrue;
    if (typeof l === "string") return { __isDicUnderConstruction: true, [l]: evalExpr(env, rExpr) };
    if (l?.__isDicUnderConstruction) {
      const key = l.__tmpKey;
      delete l.__tmpKey;
      return { ...l, [key]: evalExpr(env, rExpr) };
    }
  };
  rootEnv["=>"] = (env, l, rExpr) => (
    { env, argNames: (Array.isArray(l) ? l : [l]), expr: rExpr }
  );
  rootEnv["range"] = (env, l, rExpr) => [...Array(evalExpr(env, rExpr))].map((_, i) => i);
  rootEnv["length"] = (env, l, rExpr) => evalExpr(env, rExpr).length;
  rootEnv["map"] = (env, l, rExpr) => {
    const func = evalExpr(env, rExpr);
    return l.map((v) => callUserFunc(func, undefined, v));
  };
  rootEnv["forEach"] = rootEnv["map"];
  rootEnv["print"] = (env, l, rExpr) => log(evalExpr(env, rExpr));
  return rootEnv;
}

globalThis.log = console.log;

export function evalCode(code) {
  const tokens = tokenize(code);
  const expr = parse(tokens);
  return evalExpr(createRootEnv(), expr);
}

File: cpp/seqalt.cpp
#include <cassert>
#include <charconv>
#include <iostream>
#include <regex>
#include <string>
#include <string_view>
#include "value.hpp"

using namespace std::literals;

struct Token {
  enum class Type { SequenceStart, SequenceEnd, Number, String, Symbol };
  Type type{};
  std::string string;
};

auto tokenize(std::string_view str) {
  std::vector<Token> tokens;
  for (size_t i = 0; i < str.length();) {
    std::cmatch m;
    if (str[i] == '(' || str[i] == '{' || str[i] == '[') {
      tokens.push_back(Token{Token::Type::SequenceStart, std::string(1, str[i++])});
    } else if (str[i] == ')' || str[i] == '}' || str[i] == ']') {
      tokens.push_back(Token{Token::Type::SequenceEnd, std::string(1, str[i++])});
    } else if (std::regex_search(
      str.begin() + i, str.end(), m, std::regex("^\\d+"))
    ) {
      tokens.push_back(Token{Token::Type::Number, m[0]});
      i += m.length(0);
    } else if (std::regex_search(
      str.begin() + i, str.end(), m, std::regex(R"""(^"((\\.|[^"])*)")"""))
    ) {
      tokens.push_back(Token{
        Token::Type::String,
        std::regex_replace(std::string(m[1]), std::regex("\\\\(.)"), "$1")}
      );
      i += m.length(0);
    } else if (std::regex_search(
        str.begin() + i, str.end(), m,
        std::regex(R"""(^[A-Za-z_]\w*|^[!#-'*-/:-@^`|~]+)"""))
    ) {
      tokens.push_back(Token{Token::Type::Symbol, m[0]});
      i += m.length(0);
    } else ++i;
  }
  return tokens;
}

std::pair<size_t, Value> parseSequence(
  const std::vector<Token> &tokens, size_t i, std::string subtype = "("
);

std::pair<size_t, Value> parseExpr(
  const std::vector<Token> &tokens, size_t i
) {
  const auto &token = tokens.at(i);
  if (token.type == Token::Type::SequenceStart) {
    auto result = parseSequence(tokens, i + 1, token.string);
    assert(tokens[result.first].type == Token::Type::SequenceEnd);
    return {result.first + 1, result.second};
  }
  if (token.type == Token::Type::Number) {
    double num{};
    std::from_chars(
      token.string.data(),
      token.string.data() + token.string.size(), num
    );
    return {i + 1, Value::makeDic({{"type", "Number"}, {"value", num}})};
  }
  if (token.type == Token::Type::String) {
    return {i + 1, Value::makeDic({{"type", "String"}, {"value", token.string}})};
  }
  if (token.type == Token::Type::Symbol) {
    return {i + 1, Value::makeDic({{"type", "Symbol"}, {"value", token.string}})};
  }
  return {i, {}};
}

std::pair<size_t, Value> parseSequence(
  const std::vector<Token> &tokens, size_t i, std::string subtype
) {
  Value exprs;
  while (i < tokens.size() && tokens.at(i).type != Token::Type::SequenceEnd) {
    const auto result = parseExpr(tokens, i);
    i = result.first;
    exprs.push(result.second);
  }
  return {
    i, Value::makeDic({{"type", "Sequence"}, {"subtype", subtype}, {"exprs", exprs}})
  };
}

Value parse(const std::vector<Token> &tokens) {
  return parseSequence(tokens, 0).second;
}

Value evalExpr(Value env, Value expr);

Value callUserFunc(Value func, Value l, Value r) {
  auto env = Value::makeDic({
    {"__parentEnv", func["env"]},
    {"args", Value::makeDic({{"l", l}, {"r", r}})}
  });
  if (r.length() > 0) {
    auto argNames = func["argNames"];
    for (size_t i = 0; i < argNames.length(); ++i) {
      env[argNames[i]] = r[i];
    }
  } else {
    env[func["argNames"][0]] = r;
  }
  return evalExpr(env, func["expr"]);
}

Value callFunc(Value env, Value func, Value l, Value rExpr) {
  if (auto nativeFunc = func.asNativeFunc()) {
    return (*nativeFunc)(env, l, rExpr);
  }
  return callUserFunc(func, l, evalExpr(env, rExpr));
}

Value evalSequenceExpr(Value env, Value expr) {
  if (expr["exprs"].length() == 0) {
    if (expr["subtype"] == "[") return Value::makeArray();
    if (expr["subtype"] == "{") return Value::makeDic();
    return {};
  }
  auto val = evalExpr(env, expr["exprs"][0]);
  if (expr["subtype"] == "[") val = {{val}};
  for (size_t i = 1; i < expr["exprs"].length();) {
    auto func = evalExpr(env, expr["exprs"][i++]);
    auto rExpr = (
      i < expr["exprs"].length()
      ? expr["exprs"][i++]
      : Value::makeDic({{"type", "Null"}})
    );
    val = callFunc(env, func, val, rExpr);
  }
  if (expr["subtype"] == "{") {
    if (auto t = val.find("__type"); t && *t == "DicUnderConstruction") {
      val = val["body"];
    }
  }
  return val;
}

Value envVal(Value env, std::string_view name);

Value evalExpr(Value env, Value expr) {
  if (expr["type"] == "Sequence") return evalSequenceExpr(env, expr);
  if (expr["type"] == "Number") return expr["value"];
  if (expr["type"] == "String") return expr["value"];
  if (expr["type"] == "Symbol") return envVal(env, std::string(expr["value"]));
  return {};
}

Value ownerEnv(Value env, std::string_view name) {
  return (
    (env.isNull() || env.find(name))
    ? env : ownerEnv(env["__parentEnv"], name)
  );
}

Value envVal(Value env, std::string_view name) {
  return ownerEnv(env, name)[name];
}

Value createRootEnv() {
  Value rootEnv;
  static Value* rootEnvRef{};
  rootEnvRef = &rootEnv;
  rootEnv["@"] = Value{};
  rootEnv["//"] = Value::NativeFunc([](Value env, Value l, Value rExpr) { return l; });
  rootEnv[";"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return evalExpr(env, rExpr);
  });
  rootEnv["+"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return l + evalExpr(env, rExpr);
  });
  rootEnv["-"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(double(l) - double(evalExpr(env, rExpr)));
  });
  rootEnv["*"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(double(l) * double(evalExpr(env, rExpr)));
  });
  rootEnv["=="] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l == evalExpr(env, rExpr));
  });
  rootEnv["!="] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l != evalExpr(env, rExpr));
  });
  rootEnv["<"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l < evalExpr(env, rExpr));
  });
  rootEnv["<="] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l <= evalExpr(env, rExpr));
  });
  rootEnv[">"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l > evalExpr(env, rExpr));
  });
  rootEnv[">="] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l >= evalExpr(env, rExpr));
  });
  rootEnv["var"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    auto name = evalExpr(env, rExpr);
    env[name] = Value{};
    return name;
  });
  rootEnv["="] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    if (l.length() == 2) {
      auto obj = l[0], key = l[1];
      return obj[key] = evalExpr(env, rExpr);
    }
    auto name = std::string(l);
    auto e = ownerEnv(env, name);
    if (e.isNull()) e = *rootEnvRef;
    return e[name] = evalExpr(env, rExpr);
  });
  rootEnv["."] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return l[(rExpr["type"] == "Symbol" ? rExpr["value"] : evalExpr(env, rExpr))];
  });
  rootEnv[","] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    if (auto t = l.find("__type"); t && *t == "DicUnderConstruction") {
      auto v = l.shallowCopy();
      v["tmpKey"] = evalExpr(env, rExpr);
      return v;
    }
    if (l.length() > 0) {
      auto v = l.shallowCopy();
      v.push(evalExpr(env, rExpr));
      return v;
    }
    return Value{l, evalExpr(env,rExpr)};
  });
  rootEnv["&&"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l && evalExpr(env, rExpr));
  });
  rootEnv["||"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(l || evalExpr(env, rExpr));
  });
  rootEnv["?"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value::makeDic({{"__type", "IfThenResult"}, {"isTrue", l}, {"exprIfTrue", rExpr}});
  });
  rootEnv[":"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    if (auto t = l.find("__type"); t && *t == "IfThenResult") {
      return evalExpr(env, l["isTrue"] ? l["exprIfTrue"] : rExpr);
    }
    if (auto s = l.asString()) {
      return Value::makeDic({
        {"__type", "DicUnderConstruction"},
        {"body", Value::makeDic({{*s, evalExpr(env,rExpr)}})}
      });
    }
    if (auto t = l.find("__type"); t && *t == "DicUnderConstruction") {
      auto v = l.shallowCopy();
      v["body"] = l["body"].shallowCopy();
      v["body"][l["tmpKey"]] = evalExpr(env, rExpr);
      return v;
    }
    return Value{};
  });
  rootEnv["=>"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value::makeDic({
      {"env", env},
      {"argNames", l.length() > 0 ? l : Value{{l}}},
      {"expr", rExpr}
    });
  });
  rootEnv["range"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    Value v;
    for (size_t i = 0; i < size_t(evalExpr(env, rExpr)); ++i) {
      v.push({static_cast<double>(i)});
    }
    return v;
  });
  rootEnv["length"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    return Value(double(l.length()));
  });
  rootEnv["map"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    Value v;
    auto func = evalExpr(env, rExpr);
    for (size_t i = 0; i < l.length(); ++i) {
      v.push(callUserFunc(func, Value{}, l[i]));
    }
    return v;
  });
  rootEnv["forEach"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    auto func = evalExpr(env, rExpr);
    for (size_t i = 0; i < l.length(); ++i) {
      callUserFunc(func, Value{}, l[i]);
    }
    return Value{};
  });
  rootEnv["print"] = Value::NativeFunc([](Value env, Value l, Value rExpr) {
    std::cout << std::string(evalExpr(env, rExpr)) << std::endl;
    return Value{};
  });
  return rootEnv;
}

Value evalCode(std::string_view str) {
  const auto tokens = tokenize(str);
  auto expr = Value{parse(tokens)};
  return evalExpr(createRootEnv(), expr);
}
